### PR TITLE
Shorts tab on channel page

### PR DESCRIPTION
**Pull request by KarateCowboy**
# Add 'Shorts' tab to the Channel page

## Pull Request Type

- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Description
Adds the Shorts tab for a channel to the Channel Page. Allows for playing via direct or invidious api

### Before:
before

#### After:

after

## Testing

1) Find a channel which has Shorts and Subscribe to it.
2) View the Channels page, then click on that channel
3) The new "Shorts" tab should be visible. Click it
4) There should be a grid with Shorts in it. 
5) Click any of the shorts. You should be taken to the Watch page and be able to watch the Short

### CODE REVIEW

**Comment by absidue** [comment]
This PR doesn't actually implement the feature request that you mention in the Related issue section. As mentioned in https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1216036544 that feature request is actually for the infinitely scrolling shorts tab, although there is an unrelated discussion towards the end of that thread that mentions the shorts tab on the channels page.
I'll remove it from the pull request description for you, so that it stays open until someone actually implements it.

File: package.json
Comment (absidue) [inline package.json:80]: I can't see any code in this pull request, that uses any of the new changes in YouTube.js 2.9.0, so it doesn't need to be upgraded as part of this PR.
```suggestion
    "youtubei.js": "^2.8.0",
```

File: src/renderer/components/ft-list-video/ft-list-video.js
Comment (absidue) [inline src/renderer/components/ft-list-video/ft-list-video.js:396]: This change can be simplified to this:
```suggestion
      if (this.data.publishedText && !this.isLive) {
```

**Comment by KarateCowboy** [comment]
Hi,

So, in issue #2476 there is quite a bit going on. 

- The opening post by a deleted user asks for a dedicated shorts tab but is ambiguous beyond that. We obviously cannot ask for clarification.
- Then @ChunkyProgrammer [speculates](https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1216036544) that the user _may_ have wanted an infinite scrolling list of some type.
- Then we see @efb4f5ff-1298-471a-8973-3d47447115dc [saying](https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1304549134) there should be a shorts tab that is just like the subscriptions tab. All shorts, from all channels
- I thought there was also a user asking for a shorts tab on the channels page. Is that what you removed?
- Finally, there is @ChunkyProgrammer [commenting](https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1304549134) that there should be a toggle to disable both shorts and streams


I see content for several stories in there:

1. A shorts tab on the channel page
2. a shorts page which includes aggregated shorts from all subscriptions
3. a streams tab on the channel page
4. an infinite scrolling loop of shorts with autoplay
5. the ability to disable shorts
6. the ability to disable streams

How about we use this PR as an MVP to get shorts, at all, out the door(item 1 on the list)? I can go and create separate stories for the rest, and consult others to get clarification for details about them. Let me know your thoughts, everyone.

**Comment by absidue** [comment]
The only thing I edited was the description of this pull request, to remove the reference to a feature request for a different feature (See the screenshot below).

Why am I so sure that the original feature request is unrelated to the shorts tab on the channel page?
That feature request was opened in August, YouTube only split the videos tab up into the videos, shorts and live tabs in October/November. Additionally the requester asked for a dedicated shorts tab and Chunky pointed out that the YouTube has a dedicated infinitely scrollable shorts tab (visit https://www.youtube.com and click shorts in the sidebar), so it's highly likely that the requester meant that.

This pull request only implements the discussion that is going on below that feature request, not the actual feature request itself. I removed your "implements #number" text, so the feature request stays open until it is implemented.

I'm sorry that my words and actions came across as attacks, that wasn't my intention. My intention was just to point out that the feature that this pull request implements, is not the one in the feature request that you linked, so I removed your link to it. This pull request is completely fine (well a few code changes are needed before it gets merged, like fixing the code conflicts) as a pull request that implements the shorts tab on the channel page.

pr-edit

**Comment by KarateCowboy** [comment]
Hi absidue

Forgive me if the verbosity of my last post made it seem like I was being defensive. I interpret what you wrote as combative. There is much to this project which I am still learning: what code does what, who is active and coding, and who is keeping track of what. Because of that I decided to be very detailed about what I observe and am thinking.  Also, I am excited and charged to help out, so there is that energy.

What you are saying about Youtube splitting pieces versus the creation of the feature request makes perfect sense. I totally agree.

I will clean up the code conflicts. There are two outstanding things I see where I need some advise:

1. What do I need to do about i18n for the features, if anything?
2. My code is pointing to a git commit, per @ChunkyProgrammer 's advice. Do I need to wait for a new release, or is there a specific version I should point to now?


After that, what is the review, test, and release process?

File: src/renderer/views/Channel/Channel.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.js:550]: This is not needed for shorts

File: src/renderer/views/Channel/Channel.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.js:555]: Should probably add a separate variable for storing the shorts continuation

File: src/renderer/views/Channel/Channel.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.js:565]: should call getShortsLocal instead

File: src/renderer/views/Channel/Channel.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.js:382]: should call getShortsInvidious

File: src/renderer/views/Channel/Channel.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.js:372]: Should probably add a separate variable for storing the shorts continuation

File: src/renderer/views/Watch/Watch.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Watch/Watch.js:629]: Did you get errors from this code so you added this check or did you want to make the if statement safer?

File: static/locales/en-US.yaml
Comment (ChunkyProgrammer) [inline static/locales/en-US.yaml:516]: Add `This channel does not currently have any shorts` below

**Review by ChunkyProgrammer — COMMENTED** [review]
1) You are missing a string for 'This channel does not currently have any shorts' in the en-US locale file. Other than that i18n looks good.
image

The translations for other languages are done through Weblate and if a translation is missing in a language, it will fallback to en-US 😄 

2) Pointing to the git branch should be fine for now (we'll be using YouTube.js in FreeTube's next release instead of yt-channel-info)

"Also, while the request mentioned the ability to hide shorts, I could find no such toggle. Perhaps I could add that on to this." sorry, I just didnt remember off the top of my head if that functionality was already implemented or not 😄 if you would like to add it, you can do it in a separate PR.

If you'd like to add the ability to load more shorts to this PR:
a continuation parameter can be added to the invidious call or you can call getChannelVideosMore with the shorts continuation 

Let me know if you have any more questions. Thanks for opening a PR!

File: src/renderer/views/Channel/Channel.js
Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js:550]: ✅

File: src/renderer/views/Channel/Channel.js
Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js:555]: ✅

File: src/renderer/views/Channel/Channel.js
Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js:565]: ✅

File: src/renderer/views/Channel/Channel.js
Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js:382]: ✅

File: src/renderer/views/Channel/Channel.js
Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js:372]: ✅

File: src/renderer/views/Watch/Watch.js
Comment (KarateCowboy) [reply src/renderer/views/Watch/Watch.js:629]: I got undefined errors. Shorts I was testing do not have any storyboards. So, the next line was passing undefined to `createStoryboardUrls`. So 'yes' to both.

File: static/locales/en-US.yaml
Comment (KarateCowboy) [reply static/locales/en-US.yaml:516]: ✅

File: src/renderer/views/Watch/Watch.js
Comment (absidue) [reply src/renderer/views/Watch/Watch.js:629]: Could you provide some of the links of the shorts that were erroring, the ones I tested (Linus Tech Tips YouTube channel) all worked, so I would like to see if they have other storyboards.

**Comment by KarateCowboy** [comment]
I updated and took all of @ChunkyProgrammer 's suggestions. It appears that additional shorts are automatically loading as I scroll down the page. Hopefully it works the same for you.

> Thanks for opening a PR!

Thank you for creating this application. And, thank you for facilitating my contribution.

**Comment by ChunkyProgrammer** [comment]
Would you be able to resolve the merge conflict as well?

**Comment by KarateCowboy** [comment]
@ChunkyProgrammer Sure thing. It looks like I missed the part where you said it is OK to point to the git commit.

@absidue Sure thing. To replicate the error, remove the extra conditional. Then subscribe to the channel [Julia Stunts](https://www.youtube.com/channel/UCyViBfYJKESvPOIFRBT2ymg). Then go to that channel via "Channels > Julia Maggio". Then click the shorts tab. Then the Short "Devil Hunters@WholeWheatPete @itsKingChris & @StellaChuu".

**Comment by absidue** [comment]
@KarateCowboy Looks like YouTube returns different storyboard data for those very short shorts, only a single batch of storyboards and they look like they need parsing differently. So for the moment your "ignore the storyboards entirely" change is alright as it means that it doesn't error.
I'll create a separate pull request to fix them properly, as I still need to work out how to handle that different storyboard data.

File: src/renderer/views/Watch/Watch.js
Comment (PikachuEXE) [reply src/renderer/views/Watch/Watch.js:645]: I already don't understand the `result.storyboards?.type === 'PlayerStoryboardSpec'` here
With the new `result.storyboards.boards.length >= 3` it's even harder to understand, add some code comments?

File: src/renderer/views/Channel/Channel.js
Comment (PikachuEXE) [inline src/renderer/views/Channel/Channel.js:568]: This looks really strange (even it's valid)
```suggestion
      })
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

**Comment by PikachuEXE** [comment]
What do I do to test this?
Coz I have no idea what's the intended behaviour
Please update description so any new comer can see that too

File: src/renderer/views/Channel/Channel.js
Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js:568]: I don't know the significance of that line selecting the storyboard at index = 2 because I am new to the project. I just know that when given a Short the function was receiving undefined. I would not have touched it if it were not throwing errors.

File: src/renderer/views/Watch/Watch.js
Comment (PikachuEXE) [reply src/renderer/views/Watch/Watch.js:645]: I missed `result.storyboards.boards[2]` below oops
I find this project lacks code comments very much 😬

**Review by PikachuEXE — DISMISSED** [review]
Please also test channels **without** shorts

Tested channels with shorts
- https://www.youtube.com/channel/UCXuqSBlHAE6Xw-yeJA0Tunw
- https://www.youtube.com/channel/UCajiMK_CY9icRhLepS8_3ug

image


Tested channel **without** shorts
- https://www.youtube.com/channel/UCz7eke4JGlbtc11_6mmA7ew

image

**Review by PikachuEXE — APPROVED** [review]

File: src/renderer/views/Channel/Channel.js
Comment (ChunkyProgrammer) [inline src/renderer/views/Channel/Channel.js:548]: Remove `this.`
```suggestion
      invidiousAPICall(payload).then((response) => {
```

Comment (KarateCowboy) [reply src/renderer/views/Channel/Channel.js]: committed

**Review by PikachuEXE — APPROVED** [review]

**Review by ChunkyProgrammer — DISMISSED** [review]

**Review by absidue — COMMENTED** [review]
Why are the shorts tied to the video sort by?

**Comment by KarateCowboy** [comment]
> Why are the shorts tied to the video sort by?

At the time I was thinking the functionality is the same on the Shorts page as the Videos page. I'll make sure the shorts page has its own method and strip apart if needed.

**Comment by KarateCowboy** [comment]
Also going to remove some of the grossness I found in my additions.

**Comment by KarateCowboy** [comment]
> This pull request has conflicts, please resolve those before we can evaluate the pull request.

👍🏻 still under construction. Was working on it last night 🚧

**Comment by KarateCowboy** [comment]
Okey dokey, so:

I went out on a limb and added Jest so I can write tests before I write stuff. Personally I prefer the old Mocha/Chai/Sinon combo. But, everybody and their mother supports Jest these days _because Facebook__. Perhaps more pertinent: Vue3/Vite/Vue-test-utils all have good support for it. I saw in one of project pages that @ChunkyProgrammer made cards about upgrading to Vue3 and adding TypeScript. Having a good test suite will make that much easier.

The one caveat is that jest cannot mock/spyon function calls from outside a module IF the function is then called within the module.  See [here](https://medium.com/welldone-software/jest-how-to-mock-a-function-call-inside-a-module-21c05c57a39f). Hence, I brought in Sinon to do the job.

The Channeljs page has many methods which are the same except swapping playlists/videos/shortages or YT/Invidious.  I farmed the API calls for YT/Invidious out to the api helper library to observe separation of concerns. This will be useful when creating a dedicated Channels page from the left  menu, or when creating an endless scrolling Shorts page.

As it stands the Invidious api supports continuation for shorts but the YT channel library does not, so I left that. I implemented the Sorting, but it does not seem to work because of something with the API libraries.

Let me know your thoughts.

**Comment by KarateCowboy** [comment]
> Why are the shorts tied to the video sort by?

My mistake. Fixed in this latest commit.

**Comment by KarateCowboy** [comment]
Okey dokey, so:

I went out on a limb and added Jest so I can write tests before I write stuff. Personally I prefer the old Mocha/Chai/Sinon combo. But, everybody and their mother supports Jest these days because Facebook_. Perhaps more pertinent: Vue3/Vite/Vue-test-utils all have good support for it. I saw in one of project pages that @ChunkyProgrammer made cards about upgrading to Vue3 and adding TypeScript. Having a good test suite will make that much easier.

The one caveat is that jest cannot mock/spyon function calls from outside a module IF the function is then called within the module. See [here](https://medium.com/welldone-software/jest-how-to-mock-a-function-call-inside-a-module-21c05c57a39f). Hence, I brought in Sinon to do the job.

The Channeljs page has many methods which are the same except swapping playlists/videos/shortages or YT/Invidious. I farmed the API calls for YT/Invidious out to the api helper library to observe separation of concerns. This will be useful when creating a dedicated Channels page from the left menu, or when creating an endless scrolling Shorts page.

As it stands the Invidious api supports continuation for shorts but the YT channel library does not, so I left that. I implemented the Sorting, but it does not seem to work because of something with the API libraries.

Let me know your thoughts.

PS: Sorry for the whitespace changes. It looks like some functions have a space between the name and first paren, but some do not.

**Comment by absidue** [comment]
Hi
I'm definitely not against adding tests, however please do it in a separate pull request, I'm not a fan of having one pull request that does many unrelated changes, because that makes it extremely difficult to track which commit add/removed what.
I know it's your first pull request to FreeTube but may I remind you that you are allowed to do multiple, you don't have to cram all your contributions into a single pull request :).

File: src/renderer/helpers/api/index.js
Comment (absidue) [inline src/renderer/helpers/api/index.js:8]: I'm also not a fan of this change, as it will have to be reverted when #3143 is merged as YouTube.js works differently to yt-channel-info.

**Comment by KarateCowboy** [comment]
Okey dokey.... MUCH better.

No more of that noisy whitespace changing stuff.

The API calls for shorts did not exist prior to this PR, so it's not a refactor to put them in the API helpers file. 

If you like, I can pull the test & test boiler plate into a different PR. 

Please take a fresh look when you have time. It's much cleaner now.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@KarateCowboy could u please upload the screenshots direct on GitHub instead of Dropbox

**Comment by KarateCowboy** [comment]
> @KarateCowboy could u please upload the screenshots direct on GitHub instead of Dropbox

Done

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Noticed that the view count is off

**Comment by KarateCowboy** [comment]
> Noticed that the view count is off

I looked and that's actually what is coming through from yt-channel-info. if you `console.log` the response in `src/renderer/helpers/api/index.js` you can see:

console_and_view

And the same Short on YT:
yt_source

**Comment by absidue** [comment]
Could you please add the new tab to the tabInfoValues array on the channel page. That array is used so you can cycle the focus through the tabs by pressing the left and right arrow keys.

**Comment by ChunkyProgrammer** [comment]
> PS: This YouTube.js stuff --- how close to arrival is it?

Once https://github.com/LuanRT/YouTube.js/pull/309 gets merged then https://github.com/FreeTubeApp/FreeTube/pull/3143 will need some updates and then it can be merged

**Comment by KarateCowboy** [comment]
> > PS: This YouTube.js stuff --- how close to arrival is it?
> 
> The current plan: Once [LuanRT/YouTube.js#309](https://github.com/LuanRT/YouTube.js/pull/309) gets merged then #3143 will need some updates + testing + review and then the youtubei.js migration will be merged when approved.

Understood. I will keep an eye on that and then update accordingly.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@KarateCowboy #3143 is merged!

**Comment by KarateCowboy** [comment]
> @KarateCowboy #3143 is merged :)

I'll get to it this weekend!

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> > @KarateCowboy #3143 is merged :)
> 
> I'll get to it this weekend!

umm this weekend @KarateCowboy ?

**Comment by KarateCowboy** [comment]
Sorry. I looked at it a little last weekend but had my time cut short because I started a new job. This weekend, however, I have more time. Even if it's my birthday. :D

Mar 11, 2023 01:11:53 efb4f5ff-1298-471a-8973-3d47447115dc ***@***.***>:

**Comment by absidue** [comment]
@KarateCowboy feel free to use #3273 as an inspiration (it adds support for the live tab).

**Comment by KarateCowboy** [comment]
pulled in latest. hacking away a little tonight

**Comment by ChunkyProgrammer** [comment]
Hi @KarateCowboy do you still plan on working on this? Note: invidious shorts should work once https://github.com/iv-org/invidious/pull/3700 is merged

**Comment by KarateCowboy** [comment]
> Hi @KarateCowboy do you still plan on working on this? Note: invidious shorts should work once [iv-org/invidious#3700](https://github.com/iv-org/invidious/pull/3700) is merged

Hey hey. Thanks for the bump. I'll hit it up over the holiday weekend. Sunday night

**Comment by absidue** [comment]
Hi @KarateCowboy
Glad to hear that you are picking this up again :). To keep this pull request small and on topic, please leave the addition of a test framework and the refactor out of this pull request. If you want to add it, please do so in a different pull request, instead of cramming multiple things into one pull request. Keeping it simple, makes it easier and faster to review.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Closing, this is implemented in https://github.com/FreeTubeApp/FreeTube/pull/3533

**Pull request closed without merging**
